### PR TITLE
ENH: add pickle support for Geography

### LINKED ISSUE #54: Support pickling Geography objects
At the moment we didn't implement this, so you will get `TypeError: cannot pickle 'spherely.Point' object`, but it would be useful to support it.

To provide a full exact roundtrip through pickling, we should maybe pickle the raw unit vector (x, y, z) of the `S2Point`, instead of going through latitude/longitude like other conversions (eg WKB), because there can always be some floating point difference in this conversion.

### PR DESCRIPTION
**Pull request by benbovy**
Closes #54.

A few comments:

- A Python tuple is used here to encode extra information that is not available or exposed from `s2geography::Geography` (i.e., the simple feature type and empty info). Alternatively, we could reuse s2geometry's Encoder / Decoder and pack everything into a single sequence of bytes, which is probably slightly more efficient, but I didn't want to mess with that and chose the easy way.

- Empty geography information is already handled internally by `s2geography::Geography::EncodeTagged` but it is not really available externally so encode it separately (that is a small amount of duplicated info, which is fine I guess).

- I thought it would be nice to also encode / decode the attached `geog_index()` if it exists, but s2geography decodes a `ShapeIndexGeography` as an `EncodedShapeIndexGeography` and I haven't figured out how we can reuse the latter (s2geography functions like predicates accept `ShapeIndexGeography`... I haven't seen yet how to convert between encoded shape index vs. shape index).

- For some reason direct equality check doesn't work with geography collections after pickle roundtrip (i.e., `roundtriped_collection == original_collection`) even though their WKT representation seem consistent. Not sure why.

### CODE REVIEW

File: tests/test_geography.py [hunk 1]
```diff
@@ -123,3 +125,41 @@ def test_equality() -> None:
...
+        spherely.collection([spherely.point(40, 50)]),
+        # empty geography
+        spherely.point(),
+        spherely.linestring(),
+        spherely.polygon(),
+    ],
+)
+def test_pickle_roundtrip(geog):
+    pickled = pickle.dumps(geog)
+    roundtriped = pickle.loads(pickled)
```
Comment (jorisvandenbossche) [inline tests/test_geography.py:158]:
```suggestion
    roundtripped = pickle.loads(pickled)
```

**Review by jorisvandenbossche — APPROVED** [review]
Nice, looks good!

**Comment by jorisvandenbossche** [comment]
> For some reason direct equality check doesn't work with geography collections after pickle roundtrip (i.e., `roundtriped_collection == original_collection`) even though their WKT representation seem consistent. Not sure why.

I have been thinking that we should add some kind of "low-level representation" of a Geography object as a kind of debug info (i.e. showing which underlying S2 objects are actually stored, what kind of loops, etc), to sometimes better understand how the object is actually represented (as the python repr using WKT is always a converted representation). Maybe something like that could to debug a case like this as well.

**Comment by jorisvandenbossche** [comment]
Some possible lead, the `dimensions` and `nshape` attributes are different:

```python
In [17]: collection = spherely.collection([spherely.point(40, 50)])

In [18]: roundtripped = pickle.loads(pickle.dumps(collection))

In [19]: collection.nshape
Out[19]: 1

In [21]: roundtripped.nshape
Out[21]: 0

In [22]: collection.dimensions
Out[22]: 0

In [23]: roundtripped.dimensions
Out[23]: -1
```

**Comment by benbovy** [comment]
Good catch! Yes this looks like a bug in s2geography, where `GeographyCollection::Decode` does not set the `num_shapes_` and `total_shapes_` internal members.

I'll open an new issue on the s2geography repo and push a temporary fix here.

**Comment by benbovy** [comment]
Added a quick and dirty fix for the collection issue (clone the instance during decode so the returned object is created via the usual `s2geography::GeographyCollection` constructor that correctly updates internal members).

>  thought it would be nice to also encode / decode the attached geog_index() if it exists

This is nice to have but not really important at this point.

Ready to merge I think.

**Comment by benbovy** [comment]
(note: it will probably be useful to later expose some `s2geography::EncodeOptions` in spherely as global options - or under context management, like suggested in #70)

**Comment by benbovy** [comment]
> I have been thinking that we should add some kind of "low-level representation" of a Geography object as a kind of debug info (i.e. showing which underlying S2 objects are actually stored, what kind of loops, etc)

Yes that would be nice. Additionally, it would be nice to have some testing helpers like `spherely.testing.assert_geography_equals(a, b)` or use the `pytest_assertrepr_compare` hook to output more introspection results when the check fails.

**Review by jorisvandenbossche — APPROVED** [review]

**Pull request merged**